support-services/src/main/scala/com/gu/support/promotions/PromotionApplicator.scala

```diff
@@ -55,6 +58,56 @@ class DiscountApplicator(discount: DiscountBenefit, config: PromotionsDiscountCo
   )
 }
 
+class IntroductoryPriceApplicator(introductoryPriceBenefit: IntroductoryPriceBenefit) extends BenefitApplicator {
+  def applyTo(subscriptionData: SubscriptionData): SubscriptionData = {
+    val subscription = subscriptionData.subscription
+    val result = introductoryPriceRatePlanData(subscriptionData).map(
+      ratePlanData =>
+        subscriptionData.copy(
+          ratePlanData = subscriptionData.ratePlanData.::(ratePlanData),
+          subscription = subscription.copy(
+            contractAcceptanceDate = subscription.contractAcceptanceDate.plusWeeks(introductoryPriceBenefit.periodLength)
+          )
+        )
+    ).getOrElse(subscriptionData)
+    result
+  }
+
+
+  def introductoryPriceRatePlanData(subscriptionData: SubscriptionData): Option[RatePlanData] = {
+    // For the moment we need to find the productRatePlanId and productRatePlanChargeId for the SixWeekly version of the
+    // Quarterly GW product rate plan where the fulfilment option is the same ie. RestOfWorld or Domestic.
+    // To genericise this for N for N or even to work with other products, we would need to have a separate product rate plan
+    // for the introductory period which can apply to all products (in the way that discounts work currently) and then change
+    // the fulfilment lambda to pick this up.
+
+    val introductoryPriceStartDate = subscriptionData.subscription.contractAcceptanceDate
+
+    for {
+      recurringRatePlan <- subscriptionData.ratePlanData.headOption
+      recurringProductRatePlanId = recurringRatePlan.ratePlan.productRatePlanId
+      productRatePlansForEnvironment <- GuardianWeekly.ratePlans.find(_._2.exists(_.id == recurringProductRatePlanId)).map(_._2)
```

Comment:
can we map to _2 first?  even better if we can name it rather than refer to it by position
```suggestion
      somethingSomething = GuardianWeekly.ratePlans.map(_._2)
      productRatePlansForEnvironment <- somethingSomething.find(_.exists(_.id == recurringProductRatePlanId))
```

Reply:
This is not the same because we are searching a structure `Map[TouchPointEnvironment, List[ProductRatePlan]]` and want to find a `TouchPointEnvironment -> List[ProductRatePlan]` so that we can search only the product rate plans for the current environment. 
If we were to map first this is the same as just doing `ratePlans.values` which will flatten all the environments into one list.

Reply:
I forgot map on a Map is non law abiding in terms of restructuring a Map into a List of pairs.
However that's shocking that it would then notice if the value type is a List and then flatten them all together!